# Devoir : Manipulation du dataset Netflix
## Pipeline de données avec pandas



Vous devez travailler sur le dataset `netflix_titles.csv` (par exemple celui de Kaggle)
en utilisant **pandas** dans un autre notebook ou dans ce fichier après l'avoir complété.

Votre objectif est de construire un mini **pipeline de données** :
- chargement du dataset,
- exploration,
- validation,
- nettoyage,
- normalisation,
- transformation,
- préparation au stockage.


## 1. Chargement du dataset

**Tâches à réaliser (en code, de votre côté) :**

1. Importer la bibliothèque `pandas`.
2. Charger le fichier `netflix_titles.csv` dans un DataFrame `df`.
3. Afficher :
   - les 5 premières lignes (`head()`),
   - les informations générales (`info()`),
   - un résumé statistique (`describe(include='all')`).

**Questions :**
- Combien y a-t-il de lignes et de colonnes ?
- Quelles sont les colonnes principales ?
- Quels types de données voyez-vous (object, int, float, datetime...)?


In [73]:
import pandas as pd
df=pd.read_csv("netflix_titles.csv")
df.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,NaN,United States,"September 25, 2021",2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,NaN,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t..."
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,"September 24, 2021",2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...
3,s4,TV Show,Jailbirds New Orleans,NaN,NaN,NaN,"September 24, 2021",2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo..."
4,s5,TV Show,Kota Factory,NaN,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...


In [74]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


In [75]:
df.describe(include='all')

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
count,8807,8807,8807,6173,7982,7976,8797,8807.000000,8803,8804,8807,8807
unique,8807,2,8807,4528,7692,748,1767,NaN,17,220,514,8775
top,s8807,Movie,Zubaan,Rajiv Chilaka,David Attenborough,United States,"January 1, 2020",NaN,TV-MA,1 Season,"Dramas, International Movies","Paranormal activity at a lush, abandoned prope..."
freq,1,6131,1,19,19,2818,109,NaN,3207,1793,362,4
mean,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2014.180198,NaN,NaN,NaN,NaN
std,NaN,NaN,NaN,NaN,NaN,NaN,NaN,8.819312,NaN,NaN,NaN,NaN
min,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1925.000000,NaN,NaN,NaN,NaN
25%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2013.000000,NaN,NaN,NaN,NaN
50%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2017.000000,NaN,NaN,NaN,NaN
75%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2019.000000,NaN,NaN,NaN,NaN


Le data frame contient 8807 lignes et 12 colonnes.



Les colonnes principales sont : 'show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added','release_year', 'rating', 'duration', 'listed_in', 'description'.


Le tableau contient 11 colonnes de type 'object' et 1 colonne de type'int64'.



## 2. Exploration et visualisation simple

**Tâches (en code à faire par vous) :**

1. Calculer la répartition de la colonne `type` (Movie vs TV Show).
2. Identifier les colonnes qui contiennent le plus de valeurs manquantes
   (par exemple en calculant le pourcentage de `NaN` par colonne).
3. Afficher les 10 années de sortie (`release_year`) les plus fréquentes.

**Questions :**
- Y a-t-il plus de films ou de séries ?
- Quelles colonnes semblent les moins complètes (beaucoup de valeurs manquantes) ?
- L'offre de contenus semble-t-elle plutôt récente ou ancienne ? Justifiez avec `release_year`.


In [76]:
#1-Répartition de la colonne type
df['type'].value_counts()



,count
type,
Movie,6131
TV Show,2676


La répartition révèle qu'il y a plus de films que de série

In [77]:
#2-Identification des colonnes avec le plus de valeur manquantes
df.isna().sum()

,0
show_id,0
type,0
title,0
director,2634
cast,825
country,831
date_added,10
release_year,0
rating,4
duration,3


Les colonnes qui contiennent le plus de valeur manquantes sont: 'director','cast' et 'country'

In [78]:
df['release_year'].value_counts().head(10)

,count
release_year,
2018,1147
2017,1032
2019,1030
2020,953
2016,902
2021,592
2015,560
2014,352
2013,288


Parmi les années de sortie récentes, nous avons les années 2020,2021. Nous pouvons donc dire que l'offre de contenue est récente

## 3. Validation des données

On souhaite vérifier la **qualité minimale** du DataFrame.

**Tâches à faire en code :**

1. Vérifier que `show_id` est **unique** (pas de doublons).
2. Vérifier que la colonne `type` ne contient que les valeurs `"Movie"` et `"TV Show"`.
3. Vérifier qu'il n'existe pas d'années de sortie (`release_year`) manifestement aberrantes
   (par exemple avant 1920 ou dans le futur).

**Questions :**
- Avez-vous trouvé des doublons sur `show_id` ? Si oui, combien ? Que décidez-vous d'en faire ?
- Avez-vous trouvé des valeurs inattendues dans `type` ?
- Avez-vous trouvé des valeurs suspectes dans `release_year` ? Comment les gérez-vous ?


In [79]:
#1 Unicité de show_id
df['show_id'].is_unique

True

show_id ne contient pas de doublons

In [80]:
#2 Vérification que le type ne contient que 'Movie' et 'Tv Show'
df['type'].unique()

array(['Movie', 'TV Show'], dtype=object)

La colonne 'type' ne contient que 'Movie' et 'Tv show'

In [81]:
#3 Vérification des années de sortie abbérantes
sorted(df['release_year'].unique())

[np.int64(1925),
 np.int64(1942),
 np.int64(1943),
 np.int64(1944),
 np.int64(1945),
 np.int64(1946),
 np.int64(1947),
 np.int64(1954),
 np.int64(1955),
 np.int64(1956),
 np.int64(1958),
 np.int64(1959),
 np.int64(1960),
 np.int64(1961),
 np.int64(1962),
 np.int64(1963),
 np.int64(1964),
 np.int64(1965),
 np.int64(1966),
 np.int64(1967),
 np.int64(1968),
 np.int64(1969),
 np.int64(1970),
 np.int64(1971),
 np.int64(1972),
 np.int64(1973),
 np.int64(1974),
 np.int64(1975),
 np.int64(1976),
 np.int64(1977),
 np.int64(1978),
 np.int64(1979),
 np.int64(1980),
 np.int64(1981),
 np.int64(1982),
 np.int64(1983),
 np.int64(1984),
 np.int64(1985),
 np.int64(1986),
 np.int64(1987),
 np.int64(1988),
 np.int64(1989),
 np.int64(1990),
 np.int64(1991),
 np.int64(1992),
 np.int64(1993),
 np.int64(1994),
 np.int64(1995),
 np.int64(1996),
 np.int64(1997),
 np.int64(1998),
 np.int64(1999),
 np.int64(2000),
 np.int64(2001),
 np.int64(2002),
 np.int64(2003),
 np.int64(2004),
 np.int64(2005),
 np.int64(2006

Nous n'avons pas d'années de sortie abbérantes

## 4. Nettoyage

L'objectif est d'obtenir un DataFrame plus propre.

**Tâches (en code) :**

1. Supprimer les lignes en doublon sur `show_id` (si nécessaire).
2. Convertir la colonne `date_added` en type datetime (`to_datetime`) avec `errors='coerce'`.
3. Gérer les valeurs manquantes pour :
   - `country` (par exemple en remplaçant `NaN` par `'Unknown'`),
   - `director` (par `'Unknown'`),
   - toute autre colonne que vous jugez importante.

**Questions :**
- Quelles décisions avez-vous prises pour les valeurs manquantes ? Pourquoi ?
- Quelles colonnes avez-vous laissées avec des `NaN` (si c'est le cas) ? Pourquoi ?


In [82]:
#2 Convertit data_added en datetime
df['date_added']=pd.to_datetime(df['date_added'],errors='coerce')

In [83]:
df['country'].fillna('Unknown', inplace=True)


/tmp/ipython-input-4158059470.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['country'].fillna('Unknown', inplace=True)


In [84]:
df['director'].fillna('Unknown', inplace=True)

/tmp/ipython-input-2517827736.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['director'].fillna('Unknown', inplace=True)


In [85]:
df['cast'].fillna('Unknown', inplace=True)

/tmp/ipython-input-382634866.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['cast'].fillna('Unknown', inplace=True)


In [86]:
df=df.dropna(axis=0)

Pour les variables manquantes les colonnes 'cast' et 'date_added' ont été traités.

Les valeurs manquantes de 'cast' ont été remplacés par 'Unknown'(elles représentaient 9% de nos données et celles de 'date_added' ont été supprimés(supression des lignes car seulement 1% des données donc non significatifs)

## 5. Normalisation

On va rendre certaines colonnes plus homogènes et plus faciles à analyser.

**Tâches (en code) :**

1. Créer deux nouvelles colonnes à partir de `date_added` :
   - `year_added` (année d'ajout sur Netflix),
   - `month_added` (mois d'ajout).
2. Analyser la colonne `duration` et en dériver :
   - une colonne `duration_minutes` (durée en minutes, pour les films),
   - une colonne `duration_seasons` (nombre de saisons, pour les séries).
3. Vérifier que ces nouvelles colonnes ont des valeurs cohérentes.

**Questions :**
- Y a-t-il des valeurs de `duration` difficiles à interpréter ? Que faites-vous dans ces cas ?
- Comment interprétez-vous la différence entre `duration_minutes` et `duration_seasons` ?


In [87]:
#1 Création de deux colonnes à partir de date_added
df['year_added']=df['date_added'].dt.year
df['month_added']=df['date_added'].dt.month


In [88]:
#2Analyse de la colonne duration
df['duration_minutes']=pd.to_numeric(df['duration'].where(df['type']=='Movie').str.split('min').str[0]).astype('Int64')
df['duration_seasons']=pd.to_numeric(df['duration'].where(df['type']=='TV Show').str.split('Season').str[0].astype('Int64'))

## 6. Transformation pour l'analyse

On souhaite produire des résultats exploitables pour un analyste (ou un dashboard).

**Tâches (en code) :**

1. Calculer :
   - le nombre de titres par `release_year` ;
   - le nombre de titres par `type` et par `release_year` ;
   - le nombre de titres par pays (`country`).
2. Pour la colonne `listed_in` (genres), créer une structure permettant de comptabiliser
   le nombre de titres par genre (pensez à séparer les genres quand il y en a plusieurs dans la même cellule).

**Questions :**
- Quelles sont les années les plus riches en contenus ?
- Quels sont les genres les plus présents ?
- Quels sont les pays les plus représentés dans le catalogue ?


In [89]:
#1- Le nombre de titres par release_year
df['release_year'].value_counts()

,count
release_year,
2018,1140
2019,1030
2017,1014
2020,953
2016,878
...,...
1961,1
1925,1
1959,1


Les années les plus riches en contenues sont 2016,2017,2018,2019 et 2020

In [90]:
#Le nombre de titres par type et release_year

df[['type','release_year']].value_counts().head(10)


type     release_year
Movie    2018            767
         2017            765
         2016            658
         2019            633
         2020            517
TV Show  2020            436
         2019            397
Movie    2015            396
TV Show  2018            373
         2021            315
Name: count, dtype: int64

In [91]:
#Le nombre de titres par country
df['country'].value_counts()

,count
country,
United States,2775
India,971
Unknown,826
United Kingdom,403
Japan,240
...,...
"United States, India, France",1
"United States, United Kingdom, Morocco",1
"United Kingdom, Australia, Canada, United States",1


Les pays les plus représentés sont : United States,India,United Kingdom et Japan

## 7. Préparation au stockage

Imaginez que vous deviez stocker ces données nettoyées pour des usages futurs.

**Tâches (conceptuelles) :**

1. Décider quelles colonnes doivent absolument être conservées dans une version "analytique" du dataset
   (par exemple pour un tableau de bord Netflix).
2. Décider si vous stockeriez ce jeu de données nettoyé plutôt :
   - en CSV,
   - dans une base SQL.
   - autre
3. Expliquer les avantages et les inconvénients de votre choix.
4.  écrire le DataFrame nettoyé dans un fichier (CSV ou Parquet).

**Questions :**
- Quel format de stockage vous semble le plus adapté dans un contexte réel ? Pourquoi ?
- Comment un outil de BI (Tableau, Power BI, etc.) pourrait se connecter à ces données ?


1- Les colonnes qui doivent etre absolument conservés sont :show_id,type,title,director,cast,country,release_year,rating,listed_in,year_added,month_added,duration_minutes,duration_seasons

2-Je choisirai un stockage en CSV. L'un des avantages du CSV est sa compatibilité(lecture par les tableurs et beaucoup de logiciels) et aussi qu'il soit facilement éditable ou lit par un bloc note.
Parmi les inconvénients nous avons le stockage en ligne(toutes les données en  lignes sont conservés en chaine de caracteres uniques et continues) et aussi la difficulté à gérer les structures de données telles que les listes, les JSON...

In [93]:
#3- Ecriture du nouveau data_frame dans un fichier csv
df.to_csv('netflix_titles_modified.csv',index=False,encoding='utf-8')


## 8. Conclusion


- Quelles ont été, selon vous, les étapes les plus importantes du pipeline (chargement, validation, nettoyage, normalisation, transformation) ?
- Quelles difficultés avez-vous rencontrées lors de la manipulation du dataset Netflix ?
- Que feriez-vous ensuite si vous deviez transformer ce travail en projet complet
  (dashboard, API, modèle de recommandation, etc.) ?

---


- Toutes les étapes étaient importantes partant du chargement jusqu'à la transformation

-La gestion des valeurs manquantes au niveau des colonnes duration_minutes et duration_seasons.

-Faire des graphiques pour les observations